codegen/type_converter.go

```diff
@@ -42,6 +43,252 @@ type PackageNameResolver interface {
 	TypePackageName(thriftFile string) (string, error)
 }
 
+func (c *TypeConverter) getGoTypeName(
+	valueType compile.TypeSpec,
+) (string, error) {
+	switch s := valueType.(type) {
+	case *compile.BoolSpec:
+		return "bool", nil
+	case *compile.I8Spec:
+		return "int8", nil
+	case *compile.I16Spec:
+		return "int16", nil
+	case *compile.I32Spec:
+		return "int32", nil
+	case *compile.I64Spec:
+		return "int64", nil
+	case *compile.DoubleSpec:
+		return "float64", nil
+	case *compile.StringSpec:
+		return "string", nil
+	case *compile.BinarySpec:
+		return "[]byte", nil
+	case *compile.MapSpec:
+		panic("Not Implemented")
+	case *compile.SetSpec:
+		panic("Not Implemented")
+	case *compile.ListSpec:
+		panic("Not Implemented")
+	case *compile.EnumSpec, *compile.StructSpec, *compile.TypedefSpec:
+		return c.getIdentifierName(s)
+	default:
+		panic(fmt.Sprintf("Unknown type (%T) %v", valueType, valueType))
+	}
+}
+
+func (c *TypeConverter) getIdentifierName(
+	fieldType compile.TypeSpec,
+) (string, error) {
+	pkgName, err := c.Helper.TypePackageName(fieldType.ThriftFile())
+	if err != nil {
+		return "", errors.Wrapf(
+			err,
+			"could not lookup fieldType when building converter for %s :",
+			fieldType.ThriftName(),
+		)
+	}
+	typeName := pkgName + "." + fieldType.ThriftName()
+	return typeName, nil
+}
+
+func (c *TypeConverter) genConverterForStruct(
+	toFieldName string,
+	toFieldType *compile.StructSpec,
+	fromFieldType compile.TypeSpec,
+	fromIdentifier string,
+	keyPrefix string,
+	indent string,
+) error {
+	toIdentifier := indent + "out." + keyPrefix
+
+	typeName, err := c.getIdentifierName(toFieldType)
+	if err != nil {
+		return err
+	}
+	subToFields := toFieldType.Fields
+
+	fromFieldStruct, ok := fromFieldType.(*compile.StructSpec)
+	if !ok {
+		return errors.Errorf(
+			"could not convert struct fields, "+
+				"incompatible type for %s :",
+			toFieldName,
+		)
+	}
+
+	line := "if " + fromIdentifier + " != nil {"
+	c.Lines = append(c.Lines, line)
+
+	line = "	" + toIdentifier + " = &" + typeName + "{}"
+	c.Lines = append(c.Lines, line)
+
+	subFromFields := fromFieldStruct.Fields
+	err = c.genStructConverter(
+		keyPrefix+".",
+		indent+"	",
+		subFromFields,
+		subToFields,
+	)
+	if err != nil {
+		return err
+	}
+
+	line = "} else {"
+	c.Lines = append(c.Lines, line)
+
+	line = "	" + toIdentifier + " = nil"
+	c.Lines = append(c.Lines, line)
+
+	line = "}"
+	c.Lines = append(c.Lines, line)
+
+	return nil
+}
+
+func (c *TypeConverter) genConverterForPrimitive(
+	toField *compile.FieldSpec,
+	toIdentifier string,
+	fromIdentifier string,
+) error {
+	var line string
+	typeName, err := c.getGoTypeName(toField.Type)
+	if err != nil {
+		return err
+	}
+
+	if toField.Required {
```

Comment:
Feels like I am missing something, why does the conversion differ based on `required` or `optional`?